File: frontend/src/features/settings/projectPlanImport.ts
import type { BatchImportRow } from '../../api/projects'

export type ProjectPlanImportRow = Omit<Required<BatchImportRow>, 'workstream'> & {
  work_area: string
}

export type ProjectPlanImportError = {
  row: number
  message: string
}

export type ProjectPlanImportResult = {
  rows: ProjectPlanImportRow[]
  errors: ProjectPlanImportError[]
  headerRow: number | null
}

type ImportField = keyof ProjectPlanImportRow

const HEADER_ALIASES: Record<string, ImportField> = {
  项目: 'project_name',
  项目名称: 'project_name',
  阶段: 'project_name',
  目标: 'project_objective',
  重点工作: 'work_area',
  关键任务: 'key_task',
  关键成果: 'key_achievement',
  评价标准: 'completion_standard',
  完成标准: 'completion_standard',
  统筹人: 'coordinator',
  负责人: 'owner',
  责任人: 'owner',
  协同: 'collaborators',
  协同人: 'collaborators',
  成员: 'collaborators',
  计划时间: 'plan_time',
  计划开始时间: 'plan_start',
  计划结束时间: 'plan_end',
  重点工作计划开始时间: 'workstream_plan_start',
  重点工作计划结束时间: 'workstream_plan_end',
  当前状态: 'status',
  状态: 'status',
  完成情况: 'status',
  问题与需协调事项: 'issue',
  问题: 'issue',
  备注: 'notes',
  序号: 'sequence',
}

const EMPTY_ROW: ProjectPlanImportRow = {
  project_name: '',
  project_objective: '',
  work_area: '',
  key_task: '',
  key_achievement: '',
  completion_standard: '',
  coordinator: '',
  owner: '',
  collaborators: '',
  plan_time: '',
  plan_start: '',
  plan_end: '',
  workstream_plan_start: '',
  workstream_plan_end: '',
  status: '',
  issue: '',
  notes: '',
  sequence: '',
}

function clean(value: string | undefined): string {
  return (value ?? '').replace(/\r/g, '').trim()
}

function titleProjectName(lines: string[], headerIndex: number): string {
  for (let index = headerIndex - 1; index >= 0; index -= 1) {
    const candidate = clean(lines[index].replace(/\t/g, ' '))
    if (!candidate) continue
    const name = candidate
      .replace(/目标与重点工作计划表$/, '')
      .replace(/重点工作计划表$/, '')
      .replace(/工作计划表$/, '')
      .trim()
    if (name && name !== candidate || /计划表$/.test(candidate)) return name
    return candidate
  }
  return ''
}

function combinePlanTime(start: string, end: string): string {
  if (start && end) return `${start}~${end}`
  return start || end
}

function findHeader(lines: string[]): { index: number; mapped: Array<ImportField | null> } | null {
  for (let index = 0; index < lines.length; index += 1) {
    const headers = lines[index].split('\t').map(clean)
    const mapped = headers.map((header) => HEADER_ALIASES[header] ?? null)
    if (mapped.includes('key_task') || mapped.includes('work_area')) return { index, mapped }
  }
  return null
}

export function parseProjectPlanImportText(text: string): ProjectPlanImportResult {
  const lines = text.split('\n').map((line) => line.trimEnd()).filter((line) => line.trim())
  const header = findHeader(lines)
  if (!header) {
    return {
      rows: [],
      errors: [{ row: 1, message: '未识别到包含“重点工作”或“关键任务”的表头' }],
      headerRow: null,
    }
  }

  const titleProject = titleProjectName(lines, header.index)
  const hasProjectColumn = header.mapped.includes('project_name')
  const hasWorkstreamColumn = header.mapped.includes('work_area')
  let currentProject = titleProject
  let currentWorkArea = ''
  const rows: ProjectPlanImportRow[] = []
  const errors: ProjectPlanImportError[] = []

  for (let index = header.index + 1; index < lines.length; index += 1) {
    const cells = lines[index].split('\t')
    const row = { ...EMPTY_ROW }
    header.mapped.forEach((field, cellIndex) => {
      if (!field) return
      row[field] = clean(cells[cellIndex])
    })

    const projectName = row.project_name || currentProject
    const workArea = row.work_area || currentWorkArea || (!hasWorkstreamColumn ? row.key_task : '')
    if (projectName) currentProject = projectName
    if (workArea) currentWorkArea = workArea
    row.project_name = projectName
    row.work_area = workArea
    row.plan_time = row.plan_time || combinePlanTime(row.plan_start, row.plan_end)
    if (!row.status) row.status = '未开始'

    const missing: string[] = []
    if (!row.project_name) missing.push(hasProjectColumn ? '项目' : '标题中的项目名')
    if (!row.work_area) missing.push('重点工作')
    if (!row.key_task) missing.push('关键任务')
    if (missing.length > 0) {
      errors.push({ row: index + 1, message: `缺少${missing.join('、')}` })
      continue
    }
    rows.push(row)
  }

  return { rows, errors, headerRow: header.index + 1 }
}

export function toBatchImportRows(rows: ProjectPlanImportRow[]): BatchImportRow[] {
  return rows.map(({ work_area, ...row }) => ({ ...row, workstream: work_area }))
}

File: bowei_ai_dashboard/app/services/project_plan_import.py
from __future__ import annotations

import re
from collections.abc import Iterable

from sqlalchemy import inspect, text
from sqlalchemy.orm import Session

from .. import crud, models, schemas
from ..domain import source_type as ST


class ProjectPlanImportValidationError(ValueError):
    def __init__(self, errors: list[dict[str, object]]):
        self.errors = errors
        super().__init__("工作计划表存在无法导入的行")


def _text(value: str | None) -> str:
    return str(value or "").strip()


def _combine_plan_time(start: str | None, end: str | None) -> str:
    start_text = _text(start)
    end_text = _text(end)
    if start_text and end_text:
        return f"{start_text}~{end_text}"
    return start_text or end_text


def _split_names(value: str | None) -> list[str]:
    return [item.strip() for item in re.split(r"[/,、\\]+", _text(value)) if item.strip()]


def _person_id(name: str, db: Session) -> int | None:
    if not name:
        return None
    person = db.query(models.Person).filter(models.Person.name == name).first()
    return person.id if person else None


def _find_workstream(project_id: int, name: str, db: Session) -> models.Task | None:
    candidates = db.query(models.Task).filter(models.Task.project_id == project_id).all()
    return next(
        (row for row in candidates if not bool(row.is_deleted) and _text(row.key_task) == name),
        None,
    )


def _find_key_task(task_id: int, title: str, db: Session) -> models.SubTask | None:
    candidates = db.query(models.SubTask).filter(models.SubTask.task_id == task_id).all()
    return next(
        (row for row in candidates if not bool(row.is_deleted) and _text(row.title) == title),
        None,
    )


def _validate_rows(rows: Iterable[schemas.BatchImportRow]) -> list[dict[str, object]]:
    errors: list[dict[str, object]] = []
    for index, row in enumerate(rows, start=1):
        project_name = _text(row.project_name)
        key_task = _text(row.key_task)
        workstream = _text(row.workstream) or key_task
        missing = []
        if not project_name:
            missing.append("项目")
        if not workstream:
            missing.append("重点工作")
        if not key_task:
            missing.append("关键任务")
        if missing:
            errors.append({"row": index, "message": f"缺少{'、'.join(missing)}"})
    return errors


def _subtask_notes(row: schemas.BatchImportRow) -> str:
    lines: list[str] = []
    collaborators = _text(row.collaborators)
    notes = _text(row.notes)
    if collaborators:
        lines.append(f"协同人：{collaborators}")
    if notes:
        lines.append(notes)
    return "\n".join(lines)


def _sync_active_lifecycle(project_id: int, db: Session) -> None:
    columns = {column["name"] for column in inspect(db.connection()).get_columns("projects")}
    if "lifecycle_status" in columns:
        db.execute(
            text("UPDATE projects SET lifecycle_status = :status WHERE id = :project_id"),
            {"status": "active", "project_id": project_id},
        )


def import_project_plan_rows(
    db: Session,
    rows: list[schemas.BatchImportRow],
    current_user: str,
) -> dict[str, int | bool]:
    errors = _validate_rows(rows)
    if errors:
        raise ProjectPlanImportValidationError(errors)

    projects_created = 0
    projects_matched = 0
    tasks_created = 0
    subtasks_created = 0
    issues_created = 0
    duplicates_skipped = 0
    project_cache: dict[str, models.Project] = {}
    task_cache: dict[tuple[int, str], models.Task] = {}

    try:
        for row in rows:
            project_name = _text(row.project_name)
            workstream_name = _text(row.workstream) or _text(row.key_task)
            key_task_name = _text(row.key_task)

            if project_name not in project_cache:
                project = db.query(models.Project).filter(models.Project.name == project_name).first()
                if project is None:
                    project = models.Project(
                        name=project_name,
                        objectives=_text(row.project_objective),
                        status="active",
                        is_active=True,
                        coordinator=_text(row.coordinator),
                        owners=_text(row.owner),
                        collaborators=_text(row.collaborators),
                    )
                    db.add(project)
                    db.flush()
                    _sync_active_lifecycle(project.id, db)
                    projects_created += 1
                    crud.log(db, current_user, "批量导入建项", "project", project.id, {}, {"name": project_name})
                else:
                    projects_matched += 1
                    if not _text(project.objectives) and _text(row.project_objective):
                        project.objectives = _text(row.project_objective)
                project_cache[project_name] = project

            project = project_cache[project_name]
            task_key = (project.id, workstream_name)
            task = task_cache.get(task_key) or _find_workstream(project.id, workstream_name, db)
            if task is None:
                task_plan_time = _combine_plan_time(
                    row.workstream_plan_start,
                    row.workstream_plan_end,
                ) or _text(row.plan_time)
                task = models.Task(
                    project_id=project.id,
                    special_project=project_name,
                    key_task=workstream_name[:200],
                    key_achievement=_text(row.key_achievement)[:200],
                    completion_standard=_text(row.completion_standard),
                    coordinator=_text(row.coordinator),
                    owner=_text(row.owner),
                    owner_id=_person_id(_text(row.owner), db),
                    collaborators=_text(row.collaborators),
                    plan_time=task_plan_time,
                    status=_text(row.status) or "未开始",
                    source_type=ST.normalize("批量导入"),
                    submitter=current_user,
                )
                db.add(task)
                db.flush()
                tasks_created += 1
                crud.log(db, current_user, "批量导入建重点工作", "task", task.id, {}, {"key_task": workstream_name})
            task_cache[task_key] = task

            if _find_key_task(task.id, key_task_name, db) is not None:
                duplicates_skipped += 1
                continue

            subtask = models.SubTask(
                task_id=task.id,
                title=key_task_name[:200],
                assignee=_text(row.owner),
                assignee_id=_person_id(_text(row.owner), db),
                collaborator_ids=[person_id for person_id in (_person_id(name, db) for name in _split_names(row.collaborators)) if person_id is not None],
                plan_time=_text(row.plan_time) or _combine_plan_time(row.plan_start, row.plan_end),
                status=_text(row.status) or "未开始",
                completion_criteria="",
                notes=_subtask_notes(row),
            )
            db.add(subtask)
            db.flush()
            subtasks_created += 1
            crud.log(db, current_user, "批量导入建关键任务", "subtask", subtask.id, {}, {"title": key_task_name})

            issue_text = _text(row.issue)
            if issue_text:
                db.add(
                    models.Issue(
                        project_id=project.id,
                        special_project=project_name,
                        related_task_id=task.id,
                        description=issue_text,
                        owner=_text(row.owner),
                        source_type=ST.normalize("批量导入"),
                        status="待处理",
                        priority="中",
                    )
                )
                issues_created += 1

        db.commit()
    except Exception:
        db.rollback()
        raise

    return {
        "ok": True,
        "projects_created": projects_created,
        "projects_matched": projects_matched,
        "tasks_created": tasks_created,
        "subtasks_created": subtasks_created,
        "issues_created": issues_created,
        "duplicates_skipped": duplicates_skipped,
        "skipped_rows": 0,
    }
